### PR TITLE
Added tests for negative values of parameter scenarios

### PR DESCRIPTION
**Pull request by aditipande96**
Signed-off-by: Aditi Pande <aditi.pande@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by aditipande96** [comment]
======== 2185 tests collected in 10.87s =====

File: tests/csm/rest/test_capacity_quota.py [hunk 1]
```diff
@@ -1279,3 +1279,270 @@ def test_41156(self):
...
+        assert response.status_code == HTTPStatus.OK, \
+                  "Status code check failed for updating iam user."
+        self.log.info("Step 5: Create bucket under above IAM user")
+        self.log.info("Verify Create bucket: %s with access key: %s and secret key: %s",
+                      self.bucket, akey, skey)
+        try:
+            bucket_created = s3_misc.create_bucket(bucket_name)
+            assert_utils.assert_false(bucket_created, "Failed to create bucket")
+        except Exception as error:
+            self.log.info("Expected exception received %s", error)
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:1489]: Try put operation on already created bucket on line 1470 < ref Step 6>

File: tests/csm/rest/test_capacity_quota.py [hunk 2]
```diff
@@ -1279,3 +1279,270 @@ def test_41156(self):
...
+        self.log.info("##### Test ended -  %s #####", test_case_name)
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:1490]: Update to False again and check if the buckets are passing

File: tests/csm/rest/test_capacity_quota.py [hunk 3]
```diff
@@ -1279,3 +1279,270 @@ def test_41156(self):
...
+                                                              verify_response=True)
+        assert result, "Verification for get set user failed."
+        self.log.info("Response : %s", resp)
+        self.log.info("Step 2: Performing IOs of any size and any number of objects"
+                      "should pass")
+        random_size = math.floor(size_for_io/objects_for_io)
+        for num in range(0, objects_for_io):
+            self.log.info("Creating an uploading object %s:", num)
+            resp = s3_misc.create_put_objects(self.obj_name, self.bucket,
+                                              akey, skey, object_size=random_size)
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:1431]: self.akey, self.skey

File: tests/csm/rest/test_capacity_quota.py [hunk 4]
```diff
@@ -1279,3 +1279,270 @@ def test_41156(self):
...
+        objects_for_io = test_cfg["objects_for_io"]
+        payload = self.csm_obj.iam_user_quota_payload(quota_type, enabled, max_size,
+                                                      max_objects)
+        result, resp = self.csm_obj.verify_get_set_user_quota(self.user_id, payload,
+                                                              verify_response=True)
+        assert result, "Verification for get set user failed."
+        self.log.info("Response : %s", resp)
+        self.log.info("Step 2: Performing IOs of any size and any number of objects"
+                      "should pass")
+        random_size = math.floor(size_for_io/objects_for_io)
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:1427]: Can u move this inside loop

File: tests/csm/rest/test_capacity_quota.py (hunk 2 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py:1490]: We had decided to remove this part from TD if i am right

File: tests/csm/rest/test_capacity_quota.py [hunk 5]
```diff
@@ -1279,3 +1279,266 @@ def test_41156(self):
...
+        quota_type = test_cfg["quota_type"]
+        enabled = test_cfg["enabled"]
+        max_objects = test_cfg["max_objects"]
+        self.log.info("Step 1: Check the total available capacity")
+        resp = self.csm_obj.get_user_capacity_usage("user", self.user_id)
+        assert resp.status_code == HTTPStatus.OK, \
+            "Status code check failed for get capacity"
+        total_cap_cluster = resp.json()["capacity"]["system"]["cluster"][0]["total"]
+        total_available = resp.json()["capacity"]["system"]["cluster"][0]["available"]
+        max_size = total_available + 1024
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py [hunk 6]
```diff
@@ -1279,3 +1279,270 @@ def test_41156(self):
...
+        self.log.info("Step 2: Perform max size number of 1MB objects")
+        for num in range(0, max_size):
+            self.log.info("Creating object number %s", num)
+            resp = s3_misc.create_put_objects(self.obj_name, self.bucket,
+                                              self.akey, self.skey,
+                                              object_size=small_size)
+            assert resp, "Put object Failed"
+        self.log.info("Maximum amount of objects should be created since max_objects"
+                      "parameter is not effective")
+        self.log.info("##### Test ended -  %s #####", test_case_name)
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py:1369]: added the step

File: tests/csm/rest/test_capacity_quota.py [hunk 7]
```diff
@@ -1279,3 +1279,270 @@ def test_41156(self):
...
+    @pytest.mark.parallel
+    @pytest.mark.tags('TEST-41970')
+    def test_41970(self):
+        """
+        Test with Login and logout in loop
+        """
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        test_cfg = self.csm_conf["test_41970"]
+        custom_time = test_cfg["custom_time"]
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 1 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py:1493]: corrected

File: tests/csm/rest/test_capacity_quota.py (hunk 3 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: corrected

File: tests/csm/rest/test_capacity_quota.py (hunk 2 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py:1509]: added scenario for Suspended False

File: tests/csm/rest/test_capacity_quota.py (hunk 4 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py [hunk 8]
```diff
@@ -1279,3 +1279,266 @@ def test_41156(self):
...
+        full 100%
+        """
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        test_cfg = self.csm_conf["test_41964"]
+        quota_type = test_cfg["quota_type"]
+        enabled = test_cfg["enabled"]
+        max_objects = test_cfg["max_objects"]
+        self.log.info("Step 1: Check the total available capacity")
+        resp = self.csm_obj.get_user_capacity_usage("user", self.user_id)
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py:1301]: Okay will call get_capacity_usage() instead of this

[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py [hunk 9]
```diff
@@ -1279,3 +1279,266 @@ def test_41156(self):
...
+        assert result, "Verification for get set user failed."  # TODO: Expected outcome not known
+        self.log.info("Response : %s", resp)
+        self.log.info("Step 3: Perform & Verify GET API to get capacity usage stats")
+        resp = self.csm_obj.get_user_capacity_usage("user", self.user_id)
+        assert resp.status_code == HTTPStatus.OK, \
+            "Status code check failed for get capacity"
+        m_size = resp.json()["capacity"]["s3"]["user"][0]["used_total"]
+        assert_utils.assert_greater_equal(m_size, total_available, "Size mismatch not found")
+
+        self.log.info("Step 3: Completely full the storage capacity full")
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py [hunk 10]
```diff
@@ -1279,3 +1279,270 @@ def test_41156(self):
...
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        self.log.info("Step 1: Perform PUT API to set valid max size and negative "
+                      "max object value")
+        test_cfg = self.csm_conf["test_41965"]
+        quota_type = test_cfg["quota_type"]
+        max_size = test_cfg["max_size"]
+        small_size = test_cfg["small_size"]
+        enabled = test_cfg["enabled"]
+        max_objects = "-1024"
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: generated randomly

File: tests/csm/rest/test_capacity_quota.py [hunk 11]
```diff
@@ -1279,3 +1279,270 @@ def test_41156(self):
...
+        """
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        self.log.info("Step 1: Perform PUT API to set valid max size and negative "
+                      "max object value")
+        test_cfg = self.csm_conf["test_41967"]
+        quota_type = test_cfg["quota_type"]
+        enabled = test_cfg["enabled"]
+        max_size = "-12288"
+        max_objects = "-1024"
```
[the remark this answers is not in the document]
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py [hunk 12]
```diff
@@ -1279,3 +1279,292 @@ def test_41156(self):
...
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        self.log.info("Step 1: Perform PUT API to set valid max size and negative "
+                      "max object value")
+        test_cfg = self.csm_conf["test_41965"]
+        quota_type = test_cfg["quota_type"]
+        max_size = test_cfg["max_size"]
+        small_size = test_cfg["small_size"]
+        enabled = test_cfg["enabled"]
+        max_objects = -(self.cryptogen.randrange(1, small_size))
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:1352]: Use self.csm_obj.random_gen.randrange(1, small_size)

File: tests/csm/rest/test_capacity_quota.py [hunk 13]
```diff
@@ -1279,3 +1279,292 @@ def test_41156(self):
...
+        """
+        Test SET and GET quota with negative max size
+        """
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        self.log.info("Step 1: Perform PUT API to set valid max size and negative "
+                      "max object value")
+        test_cfg = self.csm_conf["test_41966"]
+        quota_type = test_cfg["quota_type"]
+        max_size = "-12288"
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:1390]: Use self.csm_obj.random_gen

File: tests/csm/rest/test_capacity_quota.py [hunk 14]
```diff
@@ -1279,3 +1279,292 @@ def test_41156(self):
...
+        test_case_name = cortxlogging.get_frame()
+        self.log.info("##### Test started -  %s #####", test_case_name)
+        self.log.info("Step 1: Perform PUT API to set valid max size and negative "
+                      "max object value")
+        test_cfg = self.csm_conf["test_41967"]
+        quota_type = test_cfg["quota_type"]
+        enabled = test_cfg["enabled"]
+        size_for_io = test_cfg["size_for_io"]
+        objects_for_io = test_cfg["objects_for_io"]
+        max_size = -(self.cryptogen.randrange(1, objects_for_io))
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:1425]: Use self.csm_obj.random_gen

File: tests/csm/rest/test_capacity_quota.py [hunk 15]
```diff
@@ -1279,3 +1279,292 @@ def test_41156(self):
...
+        max_objects = -(self.cryptogen.randrange(1, max_size))
```
Comment (kachhwahadivya) [inline tests/csm/rest/test_capacity_quota.py:1426]: Use self.csm_obj.random_gen

**Review by kachhwahadivya — APPROVED** [review]

File: tests/csm/rest/test_capacity_quota.py (hunk 12 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 14 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: tests/csm/rest/test_capacity_quota.py (hunk 15 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

File: config/csm/test_rest_capacity.yaml [hunk 16]
```diff
@@ -137,3 +134,25 @@ test_41156:
   min_users: 5
   max_users: 100
   available_size: 1024
+test_41964:
```
Comment (sonalk0209) [inline config/csm/test_rest_capacity.yaml:137]: I see lot of config values are common, can we combine them instead in common params unless they are absolutely test specific? @aditipande96  @kachhwahadivya

File: tests/csm/rest/test_capacity_quota.py (hunk 13 above)
Comment (aditipande96) [reply tests/csm/rest/test_capacity_quota.py]: done

**Review by sonalk0209 — APPROVED** [review]
Approving for merge

File: config/csm/test_rest_capacity.yaml (hunk 16 above)
Comment (kachhwahadivya) [reply config/csm/test_rest_capacity.yaml:137]: We can could do for some but then if in future they change for one test and not for other we are in trouble. So common parameter we put in constants.py or csm_config.yaml and here we have different ones.

**Pull request merged**